framework/src/markers/BoundaryMarker.C

```diff
@@ -0,0 +1,65 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BoundaryMarker.h"
+
+registerMooseObject("MooseApp", BoundaryMarker);
+
+InputParameters
+BoundaryMarker::validParams()
+{
+  InputParameters params = Marker::validParams();
+  params.addClassDescription(
+      "Marks all elements with sides on a given boundary for refinement/coarsening");
+  MooseEnum marker_states = Marker::markerStates();
+
+  params.addRequiredParam<MooseEnum>(
+      "mark", marker_states, "How to mark elements adjacent to the boundary.");
+  params.addRequiredParam<BoundaryName>("next_to", "Boundary to refine elements along");
+  params.addParam<Real>("distance", 0.0, "Distance from the boundary to refine within");
+  return params;
+}
+
+BoundaryMarker::BoundaryMarker(const InputParameters & parameters)
+  : Marker(parameters),
+    _distance(getParam<Real>("distance")),
+    _bnd_elem_ids(_mesh.getBoundariesToElems()),
+    _mark(parameters.get<MooseEnum>("mark").getEnum<MarkerValue>()),
+    _boundary(_mesh.getBoundaryID(getParam<BoundaryName>("next_to")))
+{
+}
+
+Marker::MarkerValue
+BoundaryMarker::computeElementMarker()
+{
+  if (_distance == 0.0)
+  {
+    if (_mesh.isBoundaryElem(_current_elem->id(), _boundary))
+      return _mark;
+
+    return DONT_MARK;
+  }
+  else
+  {
+    const auto it = _bnd_elem_ids.find(_boundary);
+    if (it != _bnd_elem_ids.end())
+      for (const auto id : it->second)
+      {
+        const auto elem = _mesh.elemPtr(id);
+        if (elem)
```

Comment:
If you're checking for existence of an element here, you want queryElemPtr(). However - I don't think you should be checking for existence. If the map isn't valid, that's bad I think? You should have only elem ids in there that you know about in distributed mesh, no?

Reply:
Ok, if I can rely on that, that's fine